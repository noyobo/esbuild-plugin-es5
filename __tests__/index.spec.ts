import esbuild, { BuildResult } from 'esbuild';
import * as path from 'path';
import * as fse from 'fs-extra';
import { Options as SWCOptions } from '@swc/core';
import { describe, it, expect } from 'vitest';

import { es5Plugin } from '../src/index.ts';

const writeFileSync = (file: string, data: string) => {
  fse.ensureFileSync(file);
  fse.writeFileSync(file, data);
};

const emitHtml = async (files: string[]) => {
  files.forEach(file => {
    let output = file.replace(path.extname(file), '.html');
    output = path.relative(__dirname, output);
    output = path.join(__dirname, '../dist', output);

    const html = `<!doctype html>
<html lang='en'>
<head>
  <meta charset='UTF-8'>
  <meta name='viewport'
        content='width=device-width, user-scalable=no, initial-scale=1.0, maximum-scale=1.0, minimum-scale=1.0'>
  <meta http-equiv='X-UA-Compatible' content='ie=edge'>
  <title>file</title>
  <link rel="stylesheet" href="./index.css">
</head>
<body>
  <script src="./index.js"></script>
</body>
</html>
`;
    fse.ensureFileSync(output);
    fse.writeFileSync(output, html);
  });
};

const createBuild = async (files: string[], swc?: SWCOptions) => {
  files = files.map(file => {
    return path.join(__dirname, file);
  });

  await emitHtml(files);
  const result = await esbuild
    .build({
      entryPoints: files,
      plugins: [es5Plugin({ swc })],
      write: false,
      bundle: false,
      sourcemap: 'linked',
      outbase: __dirname,
      outdir: 'dist',
      target: ['es5'],
      minify: false,
      // alias: {
      //   '@swc/helpers': path.dirname(require.resolve('@swc/helpers/package.json')),
      // },
    })
    .catch((error: BuildResult) => {
      expect(error.errors.length).toBe(1);
      expect(error.errors[0].text).toContain("Expected ',', got ';'");
    });

  if (result) {
    for (const file of result.outputFiles) {
      if (file.path.endsWith('.js')) {
        expect(file.text).toMatchSnapshot();
      }
      writeFileSync(file.path, file.text);
    }
  }
};

describe('index', () => {
  describe('esbuild-plugin-es5', () => {
    it('basic', async () => {
      await createBuild(['./fixtures/basic/index.ts']);
    });

    it('react', async () => {
      await createBuild(['./fixtures/react/index.tsx'], {
        module: {
          type: 'commonjs',
        },
      });
    });

    it('react-jsx', async () => {
      await createBuild(['./fixtures/react-jsx/index.jsx'], {
        module: {
          type: 'commonjs',
        },
      });
    });

    it('async', async () => {
      await createBuild(['./fixtures/async/index.ts']);
    });

    it('error', async () => {
      await createBuild(['./fixtures/error/index.ts']);
    });
  });
});
