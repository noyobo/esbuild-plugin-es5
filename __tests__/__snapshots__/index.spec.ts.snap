// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`index > esbuild-plugin-es5 > async 1`] = `
""use strict";
import { _ as _sliced_to_array } from "@swc/helpers/_/_sliced_to_array";
import { foo } from "./foo";
import { bar } from "./bar";
Promise.all([
  foo(),
  bar()
]).then(function(param) {
  var _param = _sliced_to_array(param, 2), _$foo = _param[0], _$bar = _param[1];
  console.warn(_$foo, _$bar);
});
//# sourceMappingURL=index.js.map
"
`;

exports[`index > esbuild-plugin-es5 > basic 1`] = `
""use strict";
var a = 1;
var b = "1 + 1";
console.log(a, b);
//# sourceMappingURL=index.js.map
"
`;

exports[`index > esbuild-plugin-es5 > react 1`] = `
""use strict";
Object.defineProperty(exports, "__esModule", {
  value: true
});
var _interop_require_default = require("@swc/helpers/_/_interop_require_default");
var _react = /* @__PURE__ */ _interop_require_default._(require("react"));
var _client = require("react-dom/client");
require("./index.css");
var App = function() {
  console.warn("Hello World");
  return /* @__PURE__ */ _react.default.createElement("div", null, "Hello World");
};
var div = document.createElement("div");
document.body.appendChild(div);
var root = (0, _client.createRoot)(div);
root.render(/* @__PURE__ */ _react.default.createElement(App, null));
//# sourceMappingURL=index.js.map
"
`;

exports[`index > esbuild-plugin-es5 > react-jsx 1`] = `
""use strict";
Object.defineProperty(exports, "__esModule", {
  value: true
});
var _interop_require_default = require("@swc/helpers/_/_interop_require_default");
var _react = /* @__PURE__ */ _interop_require_default._(require("react"));
var _client = require("react-dom/client");
require("./index.css");
var App = function() {
  console.warn("Hello World");
  return /* @__PURE__ */ _react.default.createElement("div", null, "Hello World");
};
var div = document.createElement("div");
document.body.appendChild(div);
var root = (0, _client.createRoot)(div);
root.render(/* @__PURE__ */ _react.default.createElement(App, null));
//# sourceMappingURL=index.js.map
"
`;
